kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-kogito/kie-wb-common-stunner-kogito-client/src/main/java/org/kie/workbench/common/stunner/kogito/client/session/command/impl/KogitoKeyEventHandlerImpl.java

```diff
@@ -52,63 +53,79 @@
                     target -> runIfEnabled(() -> shortcutCallback.onKeyUp(null)),
-                    kogitoShortcutCallback.getOpts()));
+                    kogitoShortcutCallback.getOpts()),shortcutCallback);
         } else if (shortcutCallback instanceof KogitoKeyPress) {
-            registeredShortcutsIds.add(keyboardShortcutsApi.registerKeyPress(
+            registerShortcut(keyboardShortcutsApi.registerKeyPress(
                     asStringCodes(kogitoShortcutCallback.getKeyCombination()),
                     kogitoShortcutCallback.getLabel(),
                     target -> runIfEnabled(shortcutCallback::onKeyShortcut),
-                    kogitoShortcutCallback.getOpts()));
+                    kogitoShortcutCallback.getOpts()), shortcutCallback);
         }
 
         //Session
         else if (shortcutCallback instanceof SessionKeyShortcutCallback && ((SessionKeyShortcutCallback) shortcutCallback).getDelegate() instanceof KogitoKeyShortcutKeyDownThenUp) {
-            registeredShortcutsIds.add(keyboardShortcutsApi.registerKeyDownThenUp(
+            registerShortcut(keyboardShortcutsApi.registerKeyDownThenUp(
                     asStringCodes(kogitoShortcutCallback.getKeyCombination()),
                     kogitoShortcutCallback.getLabel(),
                     target -> runIfEnabled(shortcutCallback::onKeyShortcut),
                     target -> runIfEnabled(() -> shortcutCallback.onKeyUp(null)),
-                    kogitoShortcutCallback.getOpts()));
+                    kogitoShortcutCallback.getOpts()), shortcutCallback);
         } else if (shortcutCallback instanceof SessionKeyShortcutCallback && ((SessionKeyShortcutCallback) shortcutCallback).getDelegate() instanceof KogitoKeyPress) {
-            registeredShortcutsIds.add(keyboardShortcutsApi.registerKeyPress(
+            registerShortcut(keyboardShortcutsApi.registerKeyPress(
                     asStringCodes(kogitoShortcutCallback.getKeyCombination()),
                     kogitoShortcutCallback.getLabel(),
                     target -> runIfEnabled(shortcutCallback::onKeyShortcut),
-                    kogitoShortcutCallback.getOpts()));
+                    kogitoShortcutCallback.getOpts()), shortcutCallback);
         }
 
         //Default
         else {
-            registeredShortcutsIds.add(keyboardShortcutsApi.registerKeyPress(
+            registerShortcut(keyboardShortcutsApi.registerKeyPress(
                     asStringCodes(kogitoShortcutCallback.getKeyCombination()),
                     kogitoShortcutCallback.getLabel(),
                     target -> runIfEnabled(shortcutCallback::onKeyShortcut),
-                    KeyboardShortcutsApiOpts.DEFAULT));
+                    KeyboardShortcutsApiOpts.DEFAULT), shortcutCallback);
         }
 
         return this;
     }
 
+    private void registerShortcut(Integer registeredKey, final KeyboardControl.KeyShortcutCallback callback) {
+        registeredShortcutsIds.add(registeredKey);
```

Comment:
Does it make sense to check for duplicates? Is it possible that  we add same registeredKey to the arraylist?

Reply:
Hi @domhanak, it may be possible to register the key twice, but I believe that is how the API works, for example you can register ESC to undo selection, and is also registered to be activate it when you are editing text. Plus each time a key is registered it gets a number assigned. What I did with this change is to prevent unecesary re-registering, for example: Before everytime you selected a node, 10 ESC keys were being registered but not deregistered, now it deregisters them, now there should be only a few listeres for ESC and that number should be the same no matter how many times you select and unselect a node since the listeners get deregistered.

Reply:
Thank you @inodeman for the explanation. Much appreciated!